### ISSUE TITLE
Normalization: Intensity Stats From Training and How to Use on Testing Side

### ISSUE THREAD

**Issue by amsalien** [MIC-DKFZ/nnUNet#463]
Hi: I have been reading nnunet paper and this is what I understand.

We compute following properties for each patient/case and use properties per patient to normalize data. We use ground truth mask to compute foreground pixels for all classes to compute mean, std, percentile_99_5 and percentile_00_5. These values per patient are used to normalize data? The average value across entire patient dataset is not used on training side? If we use intensity properties per patient to normalize and how can we use normalization on test/prediction side? 

# proposed 
for p in patientList:
  mean.append(np.mean(voxels[p]))
  sd.apend(np.std(voxels[p]))
  mn.append(np.min(voxels[p]))
  mx.append(np.max(voxels[p]))
  percentile_99_5.append(np.percentile(voxels[p], 99.5))
  percentile_00_5.append(np.percentile(voxels[p], 00.5))

# use these values for training and testing
np.mean(mean), np.mean(sd) and so on...

**Comment by FabianIsensee** [comment]
Hi,
this is the normalization we use for CT images. All other modalities are normalized by simply subtracting the mean and dividing by the standard deviation, case by case (this is not the mean computed over all images!!).
What you propose would be impossible to apply to test cases because you do not have a mask you could use for foreground pixels
Best,
Fabian

**Comment by amsalien** [comment]
We can use mean values from all patients from training set and use it on both training and testing side. Please let me know.

**Comment by FabianIsensee** [comment]
You can do what you want - just make sure to compare against the nnU-Net default

**Comment by amsalien** [comment]
Can you please point me to code/example how by default CT images are preprocessed on testing side in nnunet?

**Comment by amsalien** [comment]
I looked at code base, it seems approximate region of interest is picked on test side and statistics is measured from that ROI. It seems average training properties are not applied to test images. Please let me know if my understanding is correct.

**Comment by FabianIsensee** [comment]
For test images it either:
- CT: uses the values from the training set (mean, std, percentile)
- any other modality: noramlizes each case separately by its own mean and std


https://github.com/MIC-DKFZ/nnUNet/blob/6060f5d837ade6cecf15ab3dfe2ec97d75d2418a/nnunet/inference/predict.py#L35

https://github.com/MIC-DKFZ/nnUNet/blob/6060f5d837ade6cecf15ab3dfe2ec97d75d2418a/nnunet/inference/predict.py#L104

https://github.com/MIC-DKFZ/nnUNet/blob/6060f5d837ade6cecf15ab3dfe2ec97d75d2418a/nnunet/training/network_training/nnUNetTrainer.py#L416

**Comment by amsalien** [comment]
thank you for your comments. those were very useful. closing this.
